### PR TITLE
Nomgrade

### PR DESCRIPTION
**Pull request by odyslam**
## PR Checklist

- [x] Added Tests
- [ ] Updated Documentation
- [ ] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

**Comment by odyslam** [comment]
I have been building off my previous branch all this time and not the main 🎉

**Comment by anna-carroll** [comment]
should this still be draft?

File: packages/upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+scripts/upgrade.ts printGovActions -d ethereum
+```
+
+### Output & Artifacts
+
+- Artifacts & raw output are placed in `upgrade-artifacts`, a directory that is created at the root of `upgrade`
+- Artifacts are extracted from the raw output. The raw output contains ALL the relevant information of a Nomad Upgrade (Nomgrade)
+- All the files are placed under their respective domain (e.g `upgrade-artifacts/ethereum`)
+- Raw output is stored for every command that is executed
+- Artifacts are placed in `artifacts.json` - `executeCallBatchCall`: CallData to `executeCallBatch()` on the Governance Router of a domain, after it has received the Governance Message. Abi encoded function signature and arguments
```
Comment (anna-carroll) [inline packages/upgrade/README.md:73]: all of the functionality around call batches is basically duplicated with sdk-govern. we want to maintain **one** code path for doing this.

File: packages/upgrade/legacy/package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "nomad",
+    "contracts",
+    "deploy"
+  ],
+  "license": "MIT OR Apache-2.0",
+  "dependencies": {
+    "oclif": "^3.0.1"
+  },
+  "scripts": {
+    "gen-proof": "../../scripts/accumulator-cli"
```
Comment (anna-carroll) [inline packages/upgrade/legacy/package.json:17]: why is this here

**Review by anna-carroll — COMMENTED** [review]
submitting WIP review so I can do refactor with fresh slate

**Comment by odyslam** [comment]
I agree on all points about `callBatch`. From our discussions, I misunderstood that it's handled by the deploy script so I thought it would be relevant to create it here. 

We should use `sdk-govern` for sure.

File: packages/upgrade/scripts/upgrade.ts [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+      config.core[domainName].replicas[
+        Object.keys(config.core[domainName].replicas)[0]
+      ].proxy;
+
+    const governanceRouterProxy =
+      config.core[domainName].governanceRouter.proxy;
+
+    // Set env variables required by Fork test and Upgrade
+    process.env["NOMAD_REPLICA_PROXY"] = replicaProxy;
+    process.env["NOMAD_GOV_ROUTER_PROXY"] = governanceRouterProxy;
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/upgrade/scripts/upgrade.ts]: We don't need those because the `forkTest` only tests a particular part of the upgrade process, that is the change we made to message status. 

We don't test the upgraded contracts because that should be covered by the normal unit tests, IMHO.

ForkTest should only test things that we expect to break in relation to before/after.  We should expand the forkTest tests for sure

File: packages/upgrade/scripts/upgrade.ts [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    const upgradeBeaconController =
+      config.core[domainName].upgradeBeaconController;
+
+    // Set env variables to be picked up by forge script
+    // Set beacon addresses
+    process.env["NOMAD_HOME_BEACON"] = homeBeacon;
+    process.env["NOMAD_GOVERNANCE_ROUTER_BEACON"] = governanceRouterBeacon;
+    process.env["NOMAD_BRIDGE_ROUTER_BEACON"] = bridgeRouterBeacon;
+    process.env["NOMAD_TOKEN_REGISTRY_BEACON"] = tokenRegistryBeacon;
+    process.env["NOMAD_BRIDGE_ROUTER_BEACON"] = bridgeTokenBeacon;
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/upgrade/scripts/upgrade.ts]: Because I didn't test it properly. That's my bad.

File: packages/upgrade/scripts/upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+      console.log(
+        "It expects to find local EVM-compatible RPC endpoints, that listen on an incrementing port number, starting at 8545"
+      );
+      console.log(
+        "Use multi-anvil.sh to quickly spin up multiple anvil instances with incrementing port number"
+      );
+      this.announce("RPC endpoints");
+      for (let index in networks) {
+        let port: number = 8545 + parseInt(index);
+        this.nomadConfig.rpcs[networks[index]][0] = `http://127.0.0.1:${port}`;
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/upgrade/scripts/upgrade.ts]: Yes, since we are doing a `local test`, we don't need the rpc endpoints

File: packages/upgrade/scripts/upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+export class upgradeCLI extends Command {
+  static flags = {
+    // can pass either --force or -f
+    resume: Flags.boolean({ char: "r" }),
+    test: Flags.boolean({ char: "t" }),
+    config: Flags.string({
+      required: false,
+      char: "c",
+    }),
+    domains: Flags.string({
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/upgrade/scripts/upgrade.ts]: Readme will change. It's domains cause it's an array

File: packages/upgrade/utils/multi-anvil.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+
+source $(dirname "$0")/terminal-helpers.sh
+
+count=$1
+counter=0
+announce "Welcome to Multi Anvil"
+message "Mullti Anvil will generate ${count} anvil instances"
+anvils=""
+for i in $(seq $(($count - 1))); do
```
[the remark this answers is not in the document]
Comment (odyslam) [reply packages/upgrade/utils/multi-anvil.sh:11]: Not sure waht that means :P
